File: ruby/lib/rustwright/manifest.rb
# frozen_string_literal: true

require 'json'
require 'uri'
require_relative '../rustwright'

module Rustwright
  class ManifestError < Error; end

  module Manifest
    module_function

    OPERATIONS = %w[
      goto click fill title textContent evaluate screenshot
      assertTitle assertText assertEval
    ].freeze

    CAPTURE_OPERATIONS = %w[title textContent evaluate screenshot].freeze

    def load(path)
      parsed = JSON.parse(File.read(path, encoding: Encoding::UTF_8))
      validate!(parsed)
      parsed
    rescue Errno::ENOENT, Errno::EACCES => e
      raise ManifestError, "cannot read manifest #{path}: #{e.message}"
    rescue JSON::ParserError => e
      raise ManifestError, "invalid manifest JSON: #{e.message}"
    end

    def validate!(manifest)
      object!(manifest, '$')
      keys!(manifest, %w[version cases], %w[version cases], '$')
      raise ManifestError, '$.version must be 1' unless manifest['version'] == 1

      cases = manifest['cases']
      array!(cases, '$.cases')
      raise ManifestError, '$.cases must contain at least one case' if cases.empty?

      ids = {}
      cases.each_with_index do |benchmark_case, index|
        path = "$.cases[#{index}]"
        validate_case!(benchmark_case, path)
        id = benchmark_case['id']
        raise ManifestError, "duplicate case id #{id.inspect}" if ids.key?(id)

        ids[id] = true
      end
      manifest
    end

    def validate_case!(benchmark_case, path)
      object!(benchmark_case, path)
      keys!(benchmark_case, %w[id description html url steps], %w[id steps], path)
      non_empty_string!(benchmark_case['id'], "#{path}.id")
      string!(benchmark_case['description'], "#{path}.description") if benchmark_case.key?('description')
      string!(benchmark_case['html'], "#{path}.html") if benchmark_case.key?('html')
      if benchmark_case.key?('url')
        string!(benchmark_case['url'], "#{path}.url")
        begin
          URI.parse(benchmark_case['url'])
        rescue URI::InvalidURIError => e
          raise ManifestError, "#{path}.url is not a URI reference: #{e.message}"
        end
      end

      steps = benchmark_case['steps']
      array!(steps, "#{path}.steps")
      raise ManifestError, "#{path}.steps must contain at least one step" if steps.empty?

      captures = {}
      steps.each_with_index do |step, index|
        step_path = "#{path}.steps[#{index}]"
        validate_step!(step, step_path, benchmark_case)
        next unless CAPTURE_OPERATIONS.include?(step['op'])

        capture = step['capture']
        if captures.key?(capture)
          raise ManifestError, "duplicate capture #{capture.inspect} in case #{benchmark_case['id'].inspect}"
        end
        captures[capture] = true
      end
    end
    private_class_method :validate_case!

    def validate_step!(step, path, benchmark_case)
      object!(step, path)
      non_empty_string!(step['op'], "#{path}.op") if step.key?('op')
      operation = step['op']
      raise ManifestError, "#{path}.op is required" if operation.nil?
      raise ManifestError, "#{path}.op has unknown operation #{operation.inspect}" unless OPERATIONS.include?(operation)

      case operation
      when 'goto'
        keys!(step, %w[op url useCaseHtml waitUntil], %w[op], path)
        sources = [step.key?('url'), step.key?('useCaseHtml')].count(true)
        raise ManifestError, "#{path} must contain exactly one of url or useCaseHtml" unless sources == 1
        non_empty_string!(step['url'], "#{path}.url") if step.key?('url')
        if step.key?('useCaseHtml')
          raise ManifestError, "#{path}.useCaseHtml must be true" unless step['useCaseHtml'] == true
          unless benchmark_case.key?('html')
            raise ManifestError, "#{path}.useCaseHtml requires case html"
          end
        end
        if step.key?('waitUntil')
          allowed = %w[load domcontentloaded networkidle commit]
          unless allowed.include?(step['waitUntil'])
            raise ManifestError, "#{path}.waitUntil must be one of #{allowed.join(', ')}"
          end
        end
      when 'click'
        keys!(step, %w[op selector], %w[op selector], path)
        non_empty_string!(step['selector'], "#{path}.selector")
      when 'fill'
        keys!(step, %w[op selector value], %w[op selector value], path)
        non_empty_string!(step['selector'], "#{path}.selector")
        string!(step['value'], "#{path}.value")
      when 'title'
        capture_step!(step, path, %w[op capture])
      when 'textContent'
        capture_step!(step, path, %w[op selector capture])
        non_empty_string!(step['selector'], "#{path}.selector")
      when 'evaluate'
        capture_step!(step, path, %w[op expression arg capture], %w[op expression capture])
        non_empty_string!(step['expression'], "#{path}.expression")
      when 'screenshot'
        capture_step!(step, path, %w[op capture])
      when 'assertTitle'
        assertion_predicate!(step, path, %w[op equals contains])
      when 'assertText'
        assertion_predicate!(step, path, %w[op selector equals contains], %w[op selector])
        non_empty_string!(step['selector'], "#{path}.selector")
      when 'assertEval'
        keys!(step, %w[op expression equals], %w[op expression equals], path)
        non_empty_string!(step['expression'], "#{path}.expression")
      end
    end
    private_class_method :validate_step!

    def capture_step!(step, path, allowed, required = allowed)
      keys!(step, allowed, required, path)
      non_empty_string!(step['capture'], "#{path}.capture")
    end
    private_class_method :capture_step!

    def assertion_predicate!(step, path, allowed, required = %w[op])
      keys!(step, allowed, required, path)
      predicates = [step.key?('equals'), step.key?('contains')].count(true)
      raise ManifestError, "#{path} must contain exactly one of equals or contains" unless predicates == 1

      string!(step['equals'], "#{path}.equals") if step.key?('equals')
      string!(step['contains'], "#{path}.contains") if step.key?('contains')
    end
    private_class_method :assertion_predicate!

    def keys!(value, allowed, required, path)
      unknown = value.keys - allowed
      missing = required - value.keys
      raise ManifestError, "#{path} has unknown properties: #{unknown.join(', ')}" unless unknown.empty?
      raise ManifestError, "#{path} is missing required properties: #{missing.join(', ')}" unless missing.empty?
    end
    private_class_method :keys!

    def object!(value, path)
      raise ManifestError, "#{path} must be an object" unless value.is_a?(Hash)
    end
    private_class_method :object!

    def array!(value, path)
      raise ManifestError, "#{path} must be an array" unless value.is_a?(Array)
    end
    private_class_method :array!

    def string!(value, path)
      raise ManifestError, "#{path} must be a string" unless value.is_a?(String)
    end
    private_class_method :string!

    def non_empty_string!(value, path)
      string!(value, path)
      raise ManifestError, "#{path} must not be empty" if value.empty?
    end
    private_class_method :non_empty_string!
  end
end

File: go/internal/runner/manifest.go
package runner

import (
	"encoding/json"
	"fmt"
	"net/url"
	"os"
	"sort"
	"strings"
)

type Manifest struct {
	Version int
	Cases   []Case
}

type Case struct {
	ID          string
	Description string
	HTML        string
	URL         string
	Steps       []Step
}

type Step struct {
	Op           string
	URL          string
	UseCaseHTML  bool
	WaitUntil    string
	Selector     string
	Value        string
	Capture      string
	Expression   string
	Arg          json.RawMessage
	HasArg       bool
	Equals       any
	HasEquals    bool
	EqualsString string
	Contains     string
	HasContains  bool
}

func LoadManifest(path string) (*Manifest, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read manifest: %w", err)
	}
	return ParseManifest(data)
}

func ParseManifest(data []byte) (*Manifest, error) {
	root, err := parseObject(data, "manifest")
	if err != nil {
		return nil, err
	}
	if err := fields(root, "manifest", []string{"version", "cases"}, []string{"version", "cases"}); err != nil {
		return nil, err
	}
	var version float64
	if err := json.Unmarshal(root["version"], &version); err != nil || version != 1 {
		return nil, fmt.Errorf("manifest.version: expected supported version 1")
	}
	var rawCases []json.RawMessage
	if err := json.Unmarshal(root["cases"], &rawCases); err != nil {
		return nil, fmt.Errorf("manifest.cases: expected array")
	}
	if len(rawCases) == 0 {
		return nil, fmt.Errorf("manifest.cases: expected at least one case")
	}

	manifest := &Manifest{Version: 1, Cases: make([]Case, 0, len(rawCases))}
	ids := make(map[string]struct{}, len(rawCases))
	for i, raw := range rawCases {
		parsed, err := parseCase(raw, i)
		if err != nil {
			return nil, err
		}
		if _, exists := ids[parsed.ID]; exists {
			return nil, fmt.Errorf("manifest.cases[%d].id: duplicate id %q", i, parsed.ID)
		}
		ids[parsed.ID] = struct{}{}
		manifest.Cases = append(manifest.Cases, parsed)
	}
	return manifest, nil
}

func parseCase(data []byte, index int) (Case, error) {
	context := fmt.Sprintf("manifest.cases[%d]", index)
	object, err := parseObject(data, context)
	if err != nil {
		return Case{}, err
	}
	if err := fields(object, context,
		[]string{"id", "description", "html", "url", "steps"},
		[]string{"id", "steps"}); err != nil {
		return Case{}, err
	}
	var parsed Case
	if parsed.ID, err = requiredNonemptyString(object, "id", context); err != nil {
		return Case{}, err
	}
	if parsed.Description, err = optionalString(object, "description", context); err != nil {
		return Case{}, err
	}
	if parsed.HTML, err = optionalString(object, "html", context); err != nil {
		return Case{}, err
	}
	if parsed.URL, err = optionalString(object, "url", context); err != nil {
		return Case{}, err
	}
	if _, present := object["url"]; present {
		if _, err := url.Parse(parsed.URL); err != nil {
			return Case{}, fmt.Errorf("%s.url: invalid URI reference: %v", context, err)
		}
	}
	var rawSteps []json.RawMessage
	if err := json.Unmarshal(object["steps"], &rawSteps); err != nil {
		return Case{}, fmt.Errorf("%s.steps: expected array", context)
	}
	if len(rawSteps) == 0 {
		return Case{}, fmt.Errorf("%s.steps: expected at least one step", context)
	}
	parsed.Steps = make([]Step, 0, len(rawSteps))
	captures := make(map[string]struct{})
	for i, raw := range rawSteps {
		step, err := parseStep(raw, fmt.Sprintf("%s.steps[%d]", context, i))
		if err != nil {
			return Case{}, err
		}
		if step.Capture != "" {
			if _, exists := captures[step.Capture]; exists {
				return Case{}, fmt.Errorf("%s.steps[%d].capture: duplicate capture %q", context, i, step.Capture)
			}
			captures[step.Capture] = struct{}{}
		}
		parsed.Steps = append(parsed.Steps, step)
	}
	return parsed, nil
}

func parseStep(data []byte, context string) (Step, error) {
	object, err := parseObject(data, context)
	if err != nil {
		return Step{}, err
	}
	op, err := requiredNonemptyString(object, "op", context)
	if err != nil {
		return Step{}, err
	}
	step := Step{Op: op}
	switch op {
	case "goto":
		if err := fields(object, context, []string{"op", "url", "useCaseHtml", "waitUntil"}, []string{"op"}); err != nil {
			return Step{}, err
		}
		urlRaw, hasURL := object["url"]
		useRaw, hasUse := object["useCaseHtml"]
		if hasURL == hasUse {
			return Step{}, fmt.Errorf("%s: goto requires exactly one of url or useCaseHtml", context)
		}
		if hasURL {
			if err := json.Unmarshal(urlRaw, &step.URL); err != nil || step.URL == "" {
				return Step{}, fmt.Errorf("%s.url: expected nonempty string", context)
			}
		} else if err := json.Unmarshal(useRaw, &step.UseCaseHTML); err != nil || !step.UseCaseHTML {
			return Step{}, fmt.Errorf("%s.useCaseHtml: expected true", context)
		}
		if _, present := object["waitUntil"]; present {
			if step.WaitUntil, err = requiredString(object, "waitUntil", context); err != nil {
				return Step{}, err
			}
			if step.WaitUntil != "load" && step.WaitUntil != "domcontentloaded" && step.WaitUntil != "networkidle" && step.WaitUntil != "commit" {
				return Step{}, fmt.Errorf("%s.waitUntil: unsupported value %q", context, step.WaitUntil)
			}
		}
	case "click":
		if err := exactStepFields(object, context, []string{"op", "selector"}); err != nil {
			return Step{}, err
		}
		step.Selector, err = requiredNonemptyString(object, "selector", context)
	case "fill":
		if err := exactStepFields(object, context, []string{"op", "selector", "value"}); err != nil {
			return Step{}, err
		}
		if step.Selector, err = requiredNonemptyString(object, "selector", context); err == nil {
			step.Value, err = requiredString(object, "value", context)
		}
	case "title":
		if err := exactStepFields(object, context, []string{"op", "capture"}); err != nil {
			return Step{}, err
		}
		step.Capture, err = requiredNonemptyString(object, "capture", context)
	case "textContent":
		if err := exactStepFields(object, context, []string{"op", "selector", "capture"}); err != nil {
			return Step{}, err
		}
		if step.Selector, err = requiredNonemptyString(object, "selector", context); err == nil {
			step.Capture, err = requiredNonemptyString(object, "capture", context)
		}
	case "evaluate":
		if err := fields(object, context, []string{"op", "expression", "arg", "capture"}, []string{"op", "expression", "capture"}); err != nil {
			return Step{}, err
		}
		if step.Expression, err = requiredNonemptyString(object, "expression", context); err == nil {
			step.Capture, err = requiredNonemptyString(object, "capture", context)
		}
		if raw, ok := object["arg"]; ok {
			step.HasArg = true
			step.Arg = append(json.RawMessage(nil), raw...)
		}
	case "screenshot":
		if err := exactStepFields(object, context, []string{"op", "capture"}); err != nil {
			return Step{}, err
		}
		step.Capture, err = requiredNonemptyString(object, "capture", context)
	case "assertTitle":
		if err := fields(object, context, []string{"op", "equals", "contains"}, []string{"op"}); err != nil {
			return Step{}, err
		}
		err = parseStringPredicate(object, context, &step)
	case "assertText":
		if err := fields(object, context, []string{"op", "selector", "equals", "contains"}, []string{"op", "selector"}); err != nil {
			return Step{}, err
		}
		if step.Selector, err = requiredNonemptyString(object, "selector", context); err == nil {
			err = parseStringPredicate(object, context, &step)
		}
	case "assertEval":
		if err := exactStepFields(object, context, []string{"op", "expression", "equals"}); err != nil {
			return Step{}, err
		}
		if step.Expression, err = requiredNonemptyString(object, "expression", context); err != nil {
			break
		}
		step.HasEquals = true
		if err = json.Unmarshal(object["equals"], &step.Equals); err != nil {
			err = fmt.Errorf("%s.equals: invalid JSON value", context)
		}
	default:
		return Step{}, fmt.Errorf("%s.op: unknown operation %q", context, op)
	}
	if err != nil {
		return Step{}, err
	}
	return step, nil
}

func parseStringPredicate(object map[string]json.RawMessage, context string, step *Step) error {
	_, hasEquals := object["equals"]
	_, hasContains := object["contains"]
	if hasEquals == hasContains {
		return fmt.Errorf("%s: assertion requires exactly one of equals or contains", context)
	}
	if hasEquals {
		step.HasEquals = true
		var err error
		step.EqualsString, err = requiredString(object, "equals", context)
		return err
	}
	step.HasContains = true
	var err error
	step.Contains, err = requiredString(object, "contains", context)
	return err
}

func parseObject(data []byte, context string) (map[string]json.RawMessage, error) {
	var object map[string]json.RawMessage
	if err := json.Unmarshal(data, &object); err != nil || object == nil {
		return nil, fmt.Errorf("%s: expected JSON object", context)
	}
	return object, nil
}

func fields(object map[string]json.RawMessage, context string, allowed, required []string) error {
	allow := make(map[string]struct{}, len(allowed))
	for _, key := range allowed {
		allow[key] = struct{}{}
	}
	var unknown []string
	for key := range object {
		if _, ok := allow[key]; !ok {
			unknown = append(unknown, key)
		}
	}
	if len(unknown) > 0 {
		sort.Strings(unknown)
		return fmt.Errorf("%s: unknown field(s): %s", context, strings.Join(unknown, ", "))
	}
	for _, key := range required {
		if _, ok := object[key]; !ok {
			return fmt.Errorf("%s: missing required field %q", context, key)
		}
	}
	return nil
}

func exactStepFields(object map[string]json.RawMessage, context string, required []string) error {
	return fields(object, context, required, required)
}

func requiredNonemptyString(object map[string]json.RawMessage, key, context string) (string, error) {
	value, err := requiredString(object, key, context)
	if err != nil {
		return "", err
	}
	if value == "" {
		return "", fmt.Errorf("%s.%s: expected nonempty string", context, key)
	}
	return value, nil
}

func requiredString(object map[string]json.RawMessage, key, context string) (string, error) {
	raw, ok := object[key]
	if !ok {
		return "", fmt.Errorf("%s: missing required field %q", context, key)
	}
	var value *string
	if err := json.Unmarshal(raw, &value); err != nil || value == nil {
		return "", fmt.Errorf("%s.%s: expected string", context, key)
	}
	return *value, nil
}

func optionalString(object map[string]json.RawMessage, key, context string) (string, error) {
	if _, ok := object[key]; !ok {
		return "", nil
	}
	return requiredString(object, key, context)
}
